### ISSUE TITLE
Storage overhaul

### ISSUE THREAD

**Issue by axic** [ethereum/browser-solidity#384]
- [x] ~consolidate editor/storage/session calls (#367)~
- [x] move editorSize to app (#372)
- [x] move Chrome syncing to app (#373)
- [x] add config file API (#374)
- [x] document all uses of storage / editor
- [x] group remote-import code in app (split between `storage.loadFiles`, `editor.packageFiles`, `app:loadFiles`, `app:copyOver`)
- [x] add storage key prefix (#377)
- [x] implement the files API (#400)
- [x] add editor events (replace `onSetChanged` with an event) (#403)
- [x] add storage events (`fileAdded`/`fileChanged`, `fileRemoved`, `fileRenamed`) (#400)
- [x] sync file before session switch (i.e. switching tabs and before adding a new file) (#427)
- [x] rewrite editor (#401)

(This replaces #64 - see it for some details on events/methods)

**Comment by axic** [comment]
I'm thinking to change "editor" to have three state variables:
- list of files
- sessions
- storage items

List of files contains all the available files with flags (e.g. read-only, temporary). This is used to show tabs, etc.

Sessions contains the currently loaded sessions - right now that equals to list of files, but when we would have a lot files, it probably doesn't make sense to have all of them always loaded.

Storage items only contain non-readonly/non-temporary files.

**Comment by axic** [comment]
Files
- exists(path)
- get(path) -> content
- set(path, content)
- setReadOnly(bool)
- remove(path)
- rename(path, path)
- list() -> paths
- emits events:
  - fileAdded
  - fileChanged
  - fileRemoved
  - fileRenamed

Storage
- exists(path)
- get(path) -> content
- set(path, content)
- keys() -> paths
- remove(path)
- rename(path, path)

Editor
- open(path, content)
- openReadOnly(path, content)
- get(path)
- discard(path)
- emits event: changed

Basically the editor only handles the "current" file, but it can keep a cache of sessions.

(Note, Files / Storage should emit more events at some point.)

**Comment by chriseth** [comment]
Sounds good! Just one tiny thing: I think there should be a single function call to add new content that is read only. Otherwise we have to deal with things like files being open in the editor and suddenly turning read-only.

**Comment by axic** [comment]
Actually my implementation does exactly that: get/set is a thin wrapper over stroage, while addReadOnly adds a readonly file.
